drivers: stm32 qspi flash driver enables the MemoryMapped mode

FRASTM opened the pull request:
This PR is enabling the MemoryMapped to access the qua NOR flash at the end of the initialization.
With this configuration, the NOR flash access is done with memcopy in read or write
Note that the erase is not supported but writing is possible in MemoryMapped mode.
(It is not recommended to fill the memory with 0xFFFFFFFF)

"QSPI NOR devices are often used for read-only applications in memory mapped mode. And when the program instructions are executed directly from the flash memory, this is known as Execute in Place or XIP mode."

The drivers detects the  CONFIG_STM32_MEMMAP  flag to enable the MemoryMapped mode.

Requires PR [#61082](https://github.com/zephyrproject-rtos/zephyr/pull/62349)

FRASTM committed:
drivers: flash: stm32 ospi drivers with Memorymapped mode

Enable the MemoryMapped Mode for the stm32 octoFlash driver
Configure the Flash in MemoryMapped to use in XiP mode.
With this mode the erase and write are not supported.
Address and size are given by the DTS register property.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: flash: stm32 flash driver has a Kconfig STM32_MEMMAP

This CONFIG_STM32_MEMMAP is for enabling the MemoryMapped mode
on external octo or quad spi memory.
In this case, the flash_stm32_read is done in mem map mode
the flash_stm32_erase is not available.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
After running the samples/drivers/spi_flash, reading the quad NOR flash of the stm32h747_disco_m7 in MemoryMapped mode gives the 
```
Perform test on single sector


Data read : 
000ff000  read 55
000ff001  read aa
000ff002  read 66
000ff003  read 99
000ff004  read ff
000ff005  read ff
000ff006  read ff
000ff007  read ff
```

FRASTM commented:
reabse on e9dd938821e6cd0024cf5fcc50da1d3b7300b8f4

GeorgeCGV commented on soc/arm/st_stm32/stm32h7/mpu_regions.c:
```diff
@@ -9,9 +9,11 @@
+					REGION_FLASH_ATTR(REGION_FLASH_SIZE)),
 	MPU_REGION_ENTRY("SRAM", CONFIG_SRAM_BASE_ADDRESS,
-					 REGION_RAM_ATTR(REGION_SRAM_SIZE)),
+					REGION_RAM_ATTR(REGION_SRAM_SIZE)),
+	MPU_REGION_ENTRY("QSPI", 0x90000000,
+					REGION_EXTMEM_ATTR(REGION_256M)),
```
What is the point of having it in `stm32h7.dtsi` as `RAM` and here as `EXTRAM`?

GeorgeCGV commented on dts/arm/st/h7/stm32h7.dtsi:
```diff
@@ -46,9 +46,10 @@
 		reg = <0x90000000 DT_SIZE_M(256)>;
 		zephyr,memory-region = "QSPI";
-		zephyr,memory-attr = "EXTMEM";
+		zephyr,memory-attr = "RAM";
```
`REGION_RAM_ATTR` (`RAM`) - prohibits execution, doesn't it?

GeorgeCGV replied:
Doesn't apply since #61577 merge.

FRASTM committed:
boards: arm: stm32u585 disco kit for XiP on external octoflash

Define the Device tree of the b_u585i_iot02a disco kit
to access the external NOR octo-flash in MemoryMapped mode
for XiP

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: stm32h7b3 disco kit for XiP on external octoflash

Define the Device tree of the stm32h7b3i_dk disco board
to access the external NOR octo-flash in MemoryMapped mode
for XiP
Set openocd runner for debugging.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
New version where the HAL is not changed thanks to direct access to the hqspi instance in the flash_stm32_qspi driver
The flash base address and size are directly placed in the flash-controller node of the Device Tree : reg = <0x90000000 DT_SIZE_M(64)>; (no more size property required)

Tested with the stm32h747i_disco_m7 disco board ; using a sample to demonstrate the read operation in memorymapped mode (no erase)

FRASTM committed:
soc: arm: stm32u5 device with DCACHE for writing to NOR flash

Enable the DCACHE1 in INCR burt mode to allow writing to the external
NOR octoFlash when in MemoryMapped mode

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -360,6 +361,49 @@ static bool qspi_address_is_valid(const struct device *dev, off_t addr,
+#ifdef CONFIG_STM32_XIP
+
+/* Function to return true if the quad-NOR flash is in MemoryMapped else false */
+static bool qspi_is_memorymapped(const struct device *dev)
```
According to HAL team, we should be able to use the HAL_GetStatus API.

erwango replied:
> According to HAL team, we should be able to use the HAL_GetStatus API.

Update: This won't work as `HAL_GetStatus()` is based on driver internal variable (reset when switching from MCUBoot to Zephyr application), not hw setting

FRASTM committed:
drivers: clock control of the stm32h7 skips pll1 re-init

Skip the PLL1 init if it is already running, this will avoid disabling
the PLL when running after a jump from mcuboot

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Cannot read the QSPI flash correctly
(gdb) x /8x 0x90000000
0x90000000:	0x0000f33d	0x00360000	0x00000000	0x00000000
0x90000010:	0x00000000	0x00000000	0x00000000	0x00000000
0x90000020:	0x00000000	0x00000000	0x00000000	0x00000000
0x90000030:	0x00000000	0x00000000	0x00000000	0x00000000

when the STM32CubeProgrammer gives 
0x90000000:	96F3B83D	00000000	00000400	00003B36

FRASTM commented:
tested with stm32l496g_disco : XiP is possible for the external NOR quad-flash
```
*** Booting Zephyr OS build zephyr-v3.4.0-3899-g31d51386c77d ***
I: Starting Direct-XIP bootloader
I: Image 0 Primary slot: Image not found
I: Secondary slot: version=0.0.0+0
I: Image 0 loaded from the secondary slot
I: Bootloader chainload address offset: 0x0
I: Jumping to the image slot
*** Booting Zephyr OS build zephyr-v3.4.0-3899-g31d51386c77d ***
Hello World! from external flash  stm32l496g_disco      
```

FRASTM commented:
tested with  stm32h747i_disco_m7, 
```
*** Booting Zephyr OS build zephyr-v3.4.0-3899-g31d51386c77d ***
I: Starting Direct-XIP bootloader                       
I: Image 0 Primary slot: Image not found                
I: Image 0 Secondary slot: Image not found
I: No slot to load for image 0
E: Unable to find bootable image
```

FRASTM commented:
tested with stm32f469i_disco : XiP is possible for the external NOR quad-flash  
Note that dummy cycle is 10 for this board compared to 6 for stm32l496g_disco

```
*** Booting Zephyr OS build v3.5.0-rc2-65-ged053ae2f3d8 ***
I: Starting Direct-XIP bootloader
I: Image 0 Primary slot: Image not found
I: Secondary slot: version=0.0.0+0
I: Image 0 loaded from the secondary slot                               
I: Bootloader chainload address offset: 0x0                             
I: Jumping to the image slot                                            
*** Booting Zephyr OS build v3.5.0-rc2-64-g217d0764689e ***             
Hello World! from external flash  stm32f469i_disco      
```

FRASTM committed:
drivers: flash: stm32 flash base address from the DTS node

For the flash driver, the base address is the MCU internal flash
address (usualyy 0x8000000). This PR gets the that address
from the device tree node "st,stm32-nv-flash"
instead of relying on the CONFIG_FLASH_BASE_ADDRESS
which might differ when building for another flash memory.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
samples: subsys: fs_little fs on stm32u585 disco in xip

Gives a sample to execute the little fs on external memory map
(XiP) where the lfs1 partition is in internal mcu flash
The application is built/linked/stored in the external NOR
flash on slot1 partition.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
samples: subsys: fs_little fs on stm32h7b3 disco in xip

Gives a sample to execute the little fs on external memory map
(XiP) where the lfs1 partition is in internal mcu flash

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: flash: stm32 ospi driver aborts memmap before erase/write

This change is aborting the memoryMapped mode of the octo-flash
before erasing or writing the NOR. Operations are performed in
command mode.
Reading is always performed in MemoryMapped mode (memcopy)

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: stm32 disco kit for XiP on external quadflash

Define the Device tree of the stm32l496g_disco board
or stm32h750b disco kit
to access the external NOR quad-flash in MemoryMapped mode
for XiP

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: arm: stm32h7 MPU attribute for the external Memory

Define the MPU attribute to be ATTR_MPU_IO for the
qspi region, starting at 0x90000000 of the stm32h7 serie.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
samples: subsys: fs_little fs on stm32 qspi disco in xip

Gives a sample to execute the little fs on external memory map
(XiP) where the lfs1 partition is in internal mcu flash

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: flash: stm32 qspi driver set NOR flash in MemoryMapped

Configures the external NOR Flash in MemoryMapped Mode, at the end of
the NOR flash-controller initialization.
Then reading/writing are performed in MemoryMapped mode
with memcopy (and no more with command mode).
In this mode: erasing is not supported anymore.
The flash size and address are given by the DTS <reg> property.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: flash: stm32 qspi driver in Dual Flash Mode when MemoryMapped

Configure the quad-spi in DualFlash Mode when enabling the MemoryMapped
then reading is possible with memcopy.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
The PR #68607 is now replacing this PR
